### DISCUSSION TITLE
Return a Different Mock Response For the Second Time

### DISCUSSION THREAD

**Discussion by ktran1005** [pester/Pester#2368] [Q&A] [answered]
Hi everyone, I am trying to return a different value based on number of times called for a mock. I have followed others suggestions but it does not seem to work for me. My mock looks something like this:
```powershell
        $Script:counter = 0;
        Mock az {
            if ($Script:counter -eq 0) {
                $returnValue = $true
                cmd /c exit 0
            }
            else {
                $returnValue = $false
                cmd /c exit 0
            }
            $Script:counter++
            return $returnValue
        } -ParameterFilter {"$args" -match 'group exists'}
```
I wonder that if my az command is in a `while-loop` (the script is tested), can it still work for the way I mocked above?

**Comment by fflaten** [comment, marked as answer]
If `az group exists ...` is executed in a script, then no. The script has it's own script-scope, so `$script:counter` would be null instead of 0 when executing the mock. As a result, your first run-block wouldn't be called. When you increment the variable, it changes the variable in the script file so else-block would still run.

One possible workaround would be to use `if ($null -eq $Script:counter) {` as first run check. It should work with these limitations:
- You won't be able to read the counter in your Pester-test
- If your script also has a counter-variable you'd modify that. So use a unique variable name.
- The counter will reset on next script run.

_Side note: If `az group exists ..` was called from a module function, the original code would work because the mock would run in a different session state and see the test file's script-scope._

**Reply by nohwnd** [reply]
I think it should work, the scriptblocks associated with -MockWith and -ParameterFilter and executed in the session state of the test, not in the session state of the mocked command. This way you can read values from the test (e.g. $myPath) and also write them e.g. $script:myPath (or even without $script if you are mutating a reference type, like hashtable).

```
invoke-pester -container (new-pestercontainer -scriptblock { 
    describe "a" {
        it "b" { 
            function az () { }
            $script:counter = 0;
            mock az { Write-Host $script:counter; $script:counter++ }

            az
            az
            az
            az
        }
    }
})
```
This does return 0,1,2,3 as it should. I think it worked the same way in v4 as well. 

--- 

If you execute it in a separate script e.g. `& ./myScript.ps1` then the whole mocking should not work.

**Reply by fflaten** [reply]
> If you execute it in a separate script e.g. & ./myScript.ps1 then the whole mocking should not work.

Yeah, OP wrote "the script is tested" so I assumed this:

```powershell
# azMock.ps1
$i = 0
while ($i -lt 3) {
    az group exists
    $i++
}

# azMock.tests.ps1
Describe 'd script' {
    It 'i - fails' {
        $Script:counter = 0;
        Mock az {
            # This never occurs. $script:counter starts at $null when az is called from a script
            if ($Script:counter -eq 0) {
                $returnValue = $true
                #cmd /c exit 0
            }
            # This will work as intended, but also run on first try - see comment above.
            else {
                $returnValue = $false
                #cmd /c exit 0
            }
            $Script:counter++
            return $returnValue
        } -ParameterFilter { "$args" -match 'group exists' }

        $res = & "$PSScriptRoot\azMock.ps1"
        # This fails as the else-block was always executed, returning false false false
        $res | Should -Be @($true, $false, $false)

        $script:counter | Should -Be 3
    }

    It 'workaround' {
        Mock az {
            # relies on $script:counter being undefined ($null) azMock.ps1
            if ($null -eq $Script:counter) {
                $returnValue = $true
                #cmd /c exit 0
            }
            else {
                $returnValue = $false
                #cmd /c exit 0
            }
            $Script:counter++
            return $returnValue
        } -ParameterFilter { "$args" -match 'group exists' }

        # This works
        $res = & "$PSScriptRoot\azMock.ps1"
        $res | Should -Be @($true, $false, $false)

        # The counter is reset to $null in every invocation of the script, so a second run would behave the same
        $res = & "$PSScriptRoot\azMock.ps1"
        $res | Should -Be @($true, $false, $false)
    }
}
```
